### ISSUE TITLE
5D constant padding

### ISSUE THREAD

**Issue by sbhaaf** [pytorch/pytorch#2550]
Hey, I'm not sure how to make a push request, but here is the code for 5D padding.

`/torch/nn/_functions` should get the following class:

```
class ConstantPad3d(Function):

    @staticmethod
    def forward(ctx, input, pad, value=0):
        assert input.dim() == 5, 'only 5D supported for padding'
        ctx.pad = pad
        ctx.value = value
        pad_lt, pad_rt, pad_tp, pad_bm, pad_ft, pad_bk = ctx.pad
        d = input.size(2) + pad_ft + pad_bk
        h = input.size(3) + pad_tp + pad_bm
        w = input.size(4) + pad_lt + pad_rt
        assert w > 0 and h > 0 and d > 0, 'input is too small'
        ctx.input_size = input.size()

        # crop input if necessary
        output = input.new(input.size(0), input.size(1), d, h, w).fill_(ctx.value)
        c_input = input
        if pad_ft < 0:
            c_input = c_input.narrow(2, -pad_ft, c_input.size(2) + pad_ft)
        if pad_bk < 0:
            c_input = c_input.narrow(2, 0, c_input.size(2) + pad_bk)
        if pad_tp < 0:
            c_input = c_input.narrow(3, -pad_tp, c_input.size(3) + pad_tp)
        if pad_bm < 0:
            c_input = c_input.narrow(3, 0, c_input.size(3) + pad_bm)
        if pad_lt < 0:
            c_input = c_input.narrow(4, -pad_lt, c_input.size(4) + pad_lt)
        if pad_rt < 0:
            c_input = c_input.narrow(4, 0, c_input.size(4) + pad_rt)

        # crop output if necessary
        c_output = output
        if pad_ft > 0:
            c_output = c_output.narrow(2, pad_ft, c_output.size(2) - pad_ft)
        if pad_bk > 0:
            c_output = c_output.narrow(2, 0, c_output.size(2) - pad_bk)
        if pad_tp > 0:
            c_output = c_output.narrow(3, pad_tp, c_output.size(3) - pad_tp)
        if pad_bm > 0:
            c_output = c_output.narrow(3, 0, c_output.size(3) - pad_bm)
        if pad_lt > 0:
            c_output = c_output.narrow(4, pad_lt, c_output.size(4) - pad_lt)
        if pad_rt > 0:
            c_output = c_output.narrow(4, 0, c_output.size(4) - pad_rt)
        c_output.copy_(c_input)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        pad_lt, pad_rt, pad_tp, pad_bm, pad_ft, pad_bk = ctx.pad

        grad_input = Variable(grad_output.data.new(ctx.input_size).zero_())
        grad_input_slices = [slice(0, x,) for x in ctx.input_size]

        def narrow_slice(dim, start, length):
            grad_input_slices[dim] = (slice(grad_input_slices[dim].start + start,
                                            grad_input_slices[dim].start + start + length))

        def slice_length(dim):
            return grad_input_slices[dim].stop - grad_input_slices[dim].start

        #  crop grad_input if necessary
        if pad_ft < 0:
            narrow_slice(2, -pad_ft, slice_length(2) + pad_ft)
        if pad_bk < 0:
            narrow_slice(2, 0, slice_length(2) + pad_bk)
        if pad_tp < 0:
            narrow_slice(3, -pad_tp, slice_length(3) + pad_tp)
        if pad_bm < 0:
            narrow_slice(3, 0, slice_length(3) + pad_bm)
        if pad_lt < 0:
            narrow_slice(4, -pad_lt, slice_length(4) + pad_lt)
        if pad_rt < 0:
            narrow_slice(4, 0, slice_length(4) + pad_rt)

        # crop grad_output if necessary
        cg_output = grad_output
        if pad_ft > 0:
            cg_output = cg_output.narrow(2, pad_ft, cg_output.size(2) - pad_ft)
        if pad_bk > 0:
            cg_output = cg_output.narrow(2, 0, cg_output.size(2) - pad_bk)
        if pad_tp > 0:
            cg_output = cg_output.narrow(3, pad_tp, cg_output.size(3) - pad_tp)
        if pad_bm > 0:
            cg_output = cg_output.narrow(3, 0, cg_output.size(3) - pad_bm)
        if pad_lt > 0:
            cg_output = cg_output.narrow(4, pad_lt, cg_output.size(4) - pad_lt)
        if pad_rt > 0:
            cg_output = cg_output.narrow(4, 0, cg_output.size(4) - pad_rt)
        gis = tuple(grad_input_slices)
        grad_input[gis] = cg_output

        return grad_input, None, None
```

and `torch/nn/functional.py` should be updated with:

```
from ._functions.padding import ConstantPad2d, ConstantPad3d

...

def pad(input, pad, mode='constant', value=0):
    """Pads tensor.
    Currently only 2D and 3D padding supported.
    In case of 4D input tensor pad should be in form
    (pad_l, pad_r, pad_t, pad_b ).
    In case of 5D pad should be (pleft, pright, ptop, pbottom, pfront, pback)
    Args:
        input (Variable): 4D or 5D tensor
        pad (tuple): 4-elem or 6-elem tuple
        mode: 'constant', 'reflect' or 'replicate'. Default: 'constant'
        value: fill value for 'constant' padding. Default: 0
    """
    if input.dim() == 4:
        assert len(pad) == 4, '4D tensors expect 4 values for padding'
        if mode == 'constant':
            return ConstantPad2d.apply(input, pad, value)
        elif mode == 'reflect':
            return _functions.thnn.ReflectionPad2d.apply(input, *pad)
        elif mode == 'replicate':
            return _functions.thnn.ReplicationPad2d.apply(input, *pad)
    elif input.dim() == 5:
        assert len(pad) == 6, '5D tensors expect 6 values for padding'
        if mode == 'constant':
            return ConstantPad3d.apply(input, pad, value)
        elif mode == 'reflect':
            raise NotImplementedError
        elif mode == 'replicate':
            return _functions.thnn.ReplicationPad3d.apply(input, *pad)
    else:
        raise NotImplementedError("Only 4D and 5D padding is supported for now")
```

**Comment by soumith** [comment]
n-dimensional Constant padding is now supported in `F.pad`
http://pytorch.org/docs/master/nn.html#torch.nn.functional.pad

Thanks to https://github.com/pytorch/pytorch/pull/2657
